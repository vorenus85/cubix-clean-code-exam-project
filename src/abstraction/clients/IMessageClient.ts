export interface IMessageClient {
  SendNotification(message: string): void
}
