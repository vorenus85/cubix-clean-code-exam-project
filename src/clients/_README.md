The "clients" folder may contain classes or modules responsible for interacting with external services, APIs, or other systems. These classes act as clients to external resources, handling communication and data exchange. They abstract away the details of how the application communicates with external services.
