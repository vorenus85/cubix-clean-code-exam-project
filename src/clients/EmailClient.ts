import { IMessageClient } from '../abstraction/clients/IMessageClient'

export class EmailClient implements IMessageClient {
  SendNotification(message: string): void {
    console.log(`Sending Email: ${message}`)
  }
}
