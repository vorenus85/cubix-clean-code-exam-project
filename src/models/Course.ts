import{ Student} from"./Student";

export class Course{ private students:Student[] =[];
    constructor( 
        private courseName: string, 
        private startDate: Date, 
        private lengthInWeeks: number, 
        private costInHuf: number) { }
    }