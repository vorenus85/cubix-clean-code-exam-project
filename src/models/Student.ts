import { Course } from './Course'
import { Person } from './Person'

export class Student extends Person {
  constructor(
    name: string,
    email: string,
    public registeredCourses: string[] = []
  ) {
    super(name, email)
  }
}
