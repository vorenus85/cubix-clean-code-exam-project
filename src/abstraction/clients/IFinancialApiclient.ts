import { Student } from '../../models/Student'

export interface IFinancialApiClient {
  GetIsOrderPayed(student: Student): Promise<boolean>
}
