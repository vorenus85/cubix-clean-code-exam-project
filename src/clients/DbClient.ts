import { IDbClient } from '../abstraction/clients/IDbClient'
import { Course } from '../models/Course'
import { CourseStatistic } from '../models/CourseStatistic'
import { Student } from '../models/Student'
import { delay } from '../utils/common'

export class DbClient implements IDbClient {
  async AddCourse(course: Course): Promise<void> {
    await delay(500)
    console.log('Course added to database')
  }

  async AddStudentToCourse(
    student: Student,
    courseName: string
  ): Promise<void> {
    await delay(500)
    console.log('Student added to course')
  }

  async GetCourseByName(courseName: string): Promise<Course | undefined> {
    await delay(500)

    // dummy response from database
    return new Course('Clean Code', new Date('2023-12-06'), 6, 69900)
  }

  async GetCourses(): Promise<Course[]> {
    await delay(500)
    const course1 = new Course('Clean Code', new Date('2023-12-06'), 6, 69900)
    const course2 = new Course('React', new Date('2023-12-07'), 10, 175900)
    return [course1, course2]
  }

  async GetCourseStatistics(courseName: string): Promise<CourseStatistic> {
    await delay(500)
    return new CourseStatistic('Clean Code', 6, 4, 80, new Date('2024-03'))
  }
}
