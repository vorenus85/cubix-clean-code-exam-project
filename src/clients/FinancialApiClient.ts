import { IFinancialApiClient } from '../abstraction/clients/IFinancialApiclient'
import { Student } from '../models/Student'
import { delay } from '../utils/common'

export class FinancialApiClient implements IFinancialApiClient {
  public async GetIsOrderPayed(student: Student): Promise<boolean> {
    await delay(500)
    return true
  }
}
