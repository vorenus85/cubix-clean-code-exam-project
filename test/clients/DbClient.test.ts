import { DbClient } from '../../src/clients/DbClient'
import { mock, mockReset } from 'jest-mock-extended'
import { Course } from '../../src/models/Course'
import { Student } from '../../src/models/Student'

describe('DbClient tests', () => {
  let dbClient: DbClient
  const mockedCourse = mock<Course>()
  const mockedStudent = mock<Student>()

  beforeEach(() => {
    mockReset(mockedCourse)
    mockReset(mockedStudent)
    dbClient = new DbClient()
  })

  it('Should add Course to database', async () => {
    const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation()
    // Act
    await dbClient.AddCourse(mockedCourse)

    // Assert
    expect(consoleLogSpy).toHaveBeenCalledWith('Course added to database')
  })

  it('should add Student to Course', async () => {
    const courseName = 'Clean Code'
    const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation()

    // Act
    await dbClient.AddStudentToCourse(mockedStudent, courseName)

    // Assert
    expect(consoleLogSpy).toHaveBeenCalledWith('Student added to course')
  })

  it('should get Course by courseName', async () => {
    const courseName = 'Clean Code'
    const actualResult = await dbClient.GetCourseByName(courseName)

    // Assert
    expect(actualResult).toMatchSnapshot()
  })

  it('should get all courses', async () => {
    const actualResult = await dbClient.GetCourses()

    // Assert
    expect(actualResult).toMatchSnapshot()
  })

  it('should get Course Statistic by courseName', async () => {
    const courseName = 'Clean Code'

    const actualResult = await dbClient.GetCourseStatistics(courseName)

    // Assert
    expect(actualResult).toMatchSnapshot()
  })
})
