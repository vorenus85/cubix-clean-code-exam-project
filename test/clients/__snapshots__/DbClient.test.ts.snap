// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbClient tests Happy paths should get Course Statistic by courseName 1`] = `
CourseStatistic {
  "courseName": "Clean Code",
  "lastAccessed": 2024-03-01T00:00:00.000Z,
  "lecturesCompleted": 4,
  "progress": 80,
  "totalLectures": 6,
}
`;

exports[`DbClient tests Happy paths should get Course by courseName 1`] = `
Course {
  "costInHuf": 69900,
  "courseName": "Clean Code",
  "lengthInWeeks": 6,
  "startDate": 2023-12-06T00:00:00.000Z,
  "students": [],
}
`;

exports[`DbClient tests Happy paths should get all Courses 1`] = `
[
  Course {
    "costInHuf": 69900,
    "courseName": "Clean Code",
    "lengthInWeeks": 6,
    "startDate": 2023-12-06T00:00:00.000Z,
    "students": [],
  },
  Course {
    "costInHuf": 175900,
    "courseName": "React",
    "lengthInWeeks": 10,
    "startDate": 2023-12-07T00:00:00.000Z,
    "students": [],
  },
]
`;

exports[`DbClient tests Happy paths should get all courses 1`] = `
[
  Course {
    "costInHuf": 69900,
    "courseName": "Clean Code",
    "lengthInWeeks": 6,
    "startDate": 2023-12-06T00:00:00.000Z,
    "students": [],
  },
  Course {
    "costInHuf": 175900,
    "courseName": "React",
    "lengthInWeeks": 10,
    "startDate": 2023-12-07T00:00:00.000Z,
    "students": [],
  },
]
`;

exports[`DbClient tests Happy paths should get course by name 1`] = `
Course {
  "costInHuf": 69900,
  "courseName": "Clean Code",
  "lengthInWeeks": 6,
  "startDate": 2023-12-06T00:00:00.000Z,
  "students": [],
}
`;

exports[`DbClient tests should get Course Statistic by courseName 1`] = `
CourseStatistic {
  "courseName": "Clean Code",
  "lastAccessed": 2024-03-01T00:00:00.000Z,
  "lecturesCompleted": 4,
  "progress": 80,
  "totalLectures": 6,
}
`;

exports[`DbClient tests should get Course by courseName 1`] = `
Course {
  "costInHuf": 69900,
  "courseName": "Clean Code",
  "lengthInWeeks": 6,
  "startDate": 2023-12-06T00:00:00.000Z,
  "students": [],
}
`;

exports[`DbClient tests should get all courses 1`] = `
[
  Course {
    "costInHuf": 69900,
    "courseName": "Clean Code",
    "lengthInWeeks": 6,
    "startDate": 2023-12-06T00:00:00.000Z,
    "students": [],
  },
  Course {
    "costInHuf": 175900,
    "courseName": "React",
    "lengthInWeeks": 10,
    "startDate": 2023-12-07T00:00:00.000Z,
    "students": [],
  },
]
`;

exports[`DbClient tests should get course by name 1`] = `
Course {
  "costInHuf": 69900,
  "courseName": "Clean Code",
  "lengthInWeeks": 6,
  "startDate": 2023-12-06T00:00:00.000Z,
  "students": [],
}
`;
